sycl/source/detail/kernel_program_cache.hpp

```diff
@@ -82,9 +75,9 @@ class KernelProgramCache {
     return {MKernelsPerProgramCache, MKernelsPerProgramCacheMutex};
   }
 
-  template <class Predicate> void waitUntilBuilt(Predicate Pred) const {
-    std::unique_lock<std::mutex> Lock(MBuildCVMutex);
-
+  template <class Predicate>
+  void waitUntilBuilt(std::unique_lock<std::mutex> &Lock,
```

Comment:
I suggest trying to rearrange load, store, exchange operations to fit in proper order instead of employing distinct mutex.

Implementation of this approach is to check atomic variable after `MBuildCVMutex` is locked as `lock` operation may take some time. This approach will require a distinct lambda with check operation as an argument to `waitUntilBuilt()` method.

If this approach doesn't fit, I think the distinct mutex shown in current version of the patch is as good as a solution.

Reply:
What do you mean by checking the atomic variable? (is it```load()``` or something?)
I'm not quite understand, if there is no the distinct mutex for ```BuildResult```, will this approach fix the situation when notification sends before putting another thread to waiting state?

Reply:
> What do you mean by checking the atomic variable? (is it `load()` or something?)

Yes, it's  `load()` exactly.

> I'm not quite understand, if there is no the distinct mutex for ```BuildResult```, will this approach fix the situation when notification sends before putting another thread to waiting state?


`store()` operation for state of built result precedes call to `KernelProgramCache::notifyAllBuild()`.

As far as I understand the use case is the following:
 * Thread1 is building program P1. The program is built by this thread. Right now the thread is going to store build state.
 * Thread2 is building program P2. Right now it's executing `KernelProgramCache::waitUntilBuilt()`, it's in the call to `condition_variable::wait()`.
 * Thread3 is building program P1. Right now it's executing `KernelProgramCache::waitUntilBuilt()`, it's in the call to `condition_variable::wait()`.
 * Thread4 is building program P2. The program is built by this thread. Right now the thread is executing `KernelProgramCache::notifyAllBuild()` function and `condition_variable::notify_all()` is already executed.

The deadlock is going to appear in Thread3.

Now, Thread2 and Thread3 are waked up. Thread3 now owns `KernelProgramCache::MBuildCVMutex` mutex. Thread2 is in waiting state on the mutex now. Thread3 checks the state and sees `BS_InProgress` there, unblocks the mutex. Thread1 stores build state and calls to `condition_variable::notify_all()`. Thread2 locks the mutex and check build state and sees `BS_Done`, unblocks the mutex and returns from `condition_variable::wait()`. Thread3 enters waiting state on condition variable until either a spurious wakeup or some other thread calls `condition_variable::notify_all()/notify_one()`.

Now I see, why my suggestion won't help here.

Although, I have another suggestion. Let's make your solution complete. Currently we use a common `condition_variable` for each and every build. We can use a per-result `condition_variable`. This will eliminate wakeups of threads waiting on program P1 build when some thread signals on build of P2. Though there may be a restriction on number of condition variables. This restriction will originate from OS. So we should check if there is no limit on the number of condition variables (i.e. the only resource it employs is RAM).

Reply:
I'm not sure if we can use multiple ```condition_variable```s because they are not CopyConstructible, MoveConstructible, CopyAssignable or MoveAssignable.
What I did: 
1. I created two new members of ```KernelProgramCache``` class:
```c++
using ProgramBuildCVT = std::map<ProgramCacheKeyT, std::condition_variable>;
using KernelBuildCVT = std::map<RT::PiProgram, std::map<string_class, std::condition_variable>>;
...
auto& acquireProgramConditionVariables() { return MProgramBuildCVs; }
auto& acquireKernelConditionVariables() { return MKernelBuildCVs; }
...
ProgramBuildCVT MProgramBuildCVs;
KernelBuildCVT MKernelBuildCVs;
```
2. Then in passed new functor to ```getOrBuild()``` function:

* for programs:

```c++
auto CondVarsF = [](KernelProgramCache &Cache) {
  return Cache.acquireProgramConditionVariables();
};
```

* for kernels:

```c++
auto CondVarsF = [&Program](KernelProgramCache &Cache) {
  return Cache.acquireKernelConditionVariables()[Program];
};
```
3. In ```getOrBuild()``` function I inserted the new pair to MProgramBuildCVs or MKernelBuildCVs:
```c++
CondVars(KPCache).emplace(std::piecewise_construct, std::forward_as_tuple(CacheKey), std::forward_as_tuple(std::condition_variable()));
```
and got compilation problem here because copy constructor of ```std::condition_variable``` marked as deleted. It looks like we can't create a map with ```std::condition_variable``` inside.
After that I wanted to get the right condition variable like this:
```c++
std::condition_variable &BuildCV = CondVars(KPCache).at(CacheKey);
```
as pass it to ```waitUntilBuilt()``` and ```notifyAllBuild()```.

What do you think about this? I also thought about adding new field ```std::condition_variable MBuildCV;``` to ```BuildResult``` but it looks like it doesn't work because the object of ```BuildResult``` struct is unique for each thread when we enter ```getOrBuild()``` function.

Reply:
> What do you think about this? I also thought about adding new field ```std::condition_variable MBuildCV;``` to ```BuildResult``` but it looks like it doesn't work because the object of ```BuildResult``` struct is unique for each thread when we enter ```getOrBuild()``` function.

`BuildResult` should not be unique per thread. It should only be unique per building object. Note, that insertion of `BuildResult` object may not take place really (see the section of `getOrBuild()` where the call to `Cache.emplace()` takes place). Note also, that `BuildResult` is a pointer to value which contains in the cache (which is a map, actually). Hence, you may augment constructor of `BuildResult` to initialize a `condition_variable` member variable without changing its signature.

Reply:
Ah, ok, thanks a lot! I debugged it incorrectly.